# Advanced Classes & OOP Design


## Classes Revisited

Last week, we learned about defining our own classes.

Classes have **attributes**, and **methods**, which are variables & functions associated with individual **instances**.

In [5]:
import datetime

class Car:
    def __init__(self, make, model, year):
        self.make = make 
        self.model = model 
        self.year = year
    
    def compute_age(self):
        return datetime.date.today().year - self.year 

In [6]:
car1 = Car("Honda", "Accord", 2019)
car2 = Car("Toyota", "RAV4", 2006)
# make attribute is different for the two instances because their data (i.e. attributes)
# are unique to that instance 
print(f'Car 1 make = {car1.make}\nCar 2 make = {car2.make}') 

Car 1 make = Honda
Car 2 make = Toyota


### Class Attributes

Sometimes we want to share data between all instances of a given class.

All cars have 4 wheels, so we could define a shared variable accessible to all instances of the `Car` class.

To do this, we create them within the `class` body, usually right above the `__init__`.

In [2]:
import datetime

class Car:
    # class attribute
    wheels = 4
    
    def __init__(self, make, model, year):
        self.make = make 
        self.model = model 
        self.year = year
    
    def compute_age(self):
        return datetime.date.today().year - self.year 
    
    
car1 = Car("Honda", "Accord", 2019)
car2 = Car("Toyota", "RAV4", 2006)

In [3]:
# class attribute can be accessed on instances, or the class itself

print(Car.wheels)
print(car1.wheels)
print(car2.wheels)

4
4
4


In [4]:
# these are all the same variable
Car.wheels is car1.wheels

True

In [5]:
# which means changes to the class attribute
# will modify for all classes 

Car.wheels = 3
print(car1.wheels)
print(car2.wheels)

3
3


In [6]:
# note: assigning to an instance attribute makes a new attribute

# creates a new instance variable!
car2.wheels = 2
print(car2.wheels is car1.wheels)
print(car1.wheels)
print(car2.wheels)
print(Car.wheels)

False
3
2
3


### Class Methods

It can also be useful to provide methods that are accessible to all instances of a class.

Class methods are similar to instance methods with a few distinctions:

1. They can not access instance methods or attributes.
2. The first argument to the method is not `self`, but instead `cls` by convention.  `cls` is the class object itself (e.g. `Car`)
3. Class methods are declared with the `@classmethod` decorator.

In [12]:
from datetime import date

class Car: 
    
    # wheels class attribute 
    wheels = 4
    # tire pressure class attribute  
    psi = 35 
    
    def __init__(self, make, model, year):
        self.make = make 
        self.model = model 
        self.year = year
    
    def compute_age(self):
        current_year = int(date.today().year)
        return current_year - self.year 
    
    @classmethod 
    def tire_description(cls):
        return f'Car has {cls.wheels} wheels with a tire pressure of {Car.psi}' 
    
car1 = Car("Honda", "Accord", 2019)
car2 = Car("Toyota", "RAV4", 2006)

In [13]:
print(Car.tire_description())
print(car1.tire_description())

Car has 4 wheels with a tire pressure of 35
Car has 4 wheels with a tire pressure of 35


Notice that we can use `Car.psi` or `cls.wheels` to access class attributes. `cls` is generally preferred, both to avoid repetition and for reasons we'll see when we get to inheritance.

Finally, note that we can access class methods and instances from within instance methods. (but not vice-versa!)

In [7]:
from datetime import date
class Car: 
    
    # wheels class attribute 
    wheels = 4
    
    # tire pressure amount 
    psi = 35 
    
    def __init__(self, make, model, year):
        self.make = make 
        self.model = model 
        self.year = year
    
    def compute_age(self):
        current_year = int(date.today().year)
        return current_year - self.year 
    
    @classmethod 
    def tire_description(cls):
        return f'Car has {cls.wheels} wheels, each with a tire pressure of {Car.psi}' 

    def __repr__(self): 
        instance_str = f'Car(make={self.make}, model={self.model}, year={self.year}, '
        instance_str += f'wheels={Car.wheels}, {self.tire_description()})'
        return instance_str

In [15]:
car1 = Car("Honda", "Civic", 2019)
print(car1)

Car(make=Honda, model=Civic, year=2019, wheels=4, Car has 4 wheels, each with a tire pressure of 35)


### Alternate Constructors

A common use of class methods is to define alternate ways to initialize an isntance.  In Python there can only be one constructor (`__init__`), whereas some other languages allow multiple.

Perhaps we have Car data coming from a file, meaning we'd have strings like:

In [11]:
car1str = "Pontiac|Grand Am|1997"
car2str = "Ford|Mustang|1970"
car3str = "Hyundai|Sonata|2007"

In [12]:
from datetime import date
class Car: 
    wheels = 4
    psi = 35
    
    def __init__(self, make, model, year):
        self.make = make 
        self.model = model 
        self.year = year
        
    @classmethod
    def from_string(cls, string):
        make, model, year = string.split("|")
        # invoke Car's constructor
        return cls(make, model, year)
    
    def compute_age(self):
        current_year = int(date.today().year)
        return current_year - self.year 
    
    @classmethod 
    def tire_description(cls):
        return f'Car has {cls.wheels} wheels, each with a tire pressure of {Car.psi}' 

    def __repr__(self): 
        instance_str = f'Car(make={self.make}, model={self.model}, year={self.year}, '
        instance_str += f'wheels={Car.wheels})'
        return instance_str

In [13]:
car1

In [14]:
car1 = Car.from_string(car1str)
car2 = Car.from_string(car2str)
car3 = Car.from_string(car3str)

In [15]:
print(car1)
print(car2)
print(car3)

Car(make=Pontiac, model=Grand Am, year=1997, wheels=4)
Car(make=Ford, model=Mustang, year=1970, wheels=4)
Car(make=Hyundai, model=Sonata, year=2007, wheels=4)


This is a common pattern, seen throughout Python:

 - ``int.from_bytes()``
 - ``float.fromhex()`` 
 - ``datetime.date.fromtimestamp()``
 - ``itertools.chain.from_iterable()``


### staticmethod

Sometimes it makes sense to just attach a method to a class for the purpose of namespacing.

In [20]:
def which_is_newer(a, b):
    if a.year > b.year:
        return a
    else:
        return b

which_is_newer(car1, car2)

Car(make=Pontiac, model=Grand Am, year=1997, wheels=4)

In [16]:
# it might make sense to attach this to the class, 
# but neither a classmethod nor an instance method

from datetime import date
class Car: 
    wheels = 4
    psi = 35
    
    # does not take self or cls
    @staticmethod
    def which_is_newer(a, b):
        if a.year > b.year:
            return a
        else:
            return b
        
    @staticmethod
    def something():
        return []
    

    
    def __init__(self, make, model, year):
        self.make = make 
        self.model = model 
        self.year = year
        
    @classmethod
    def from_string(cls, string):
        make, model, year = string.split("|")
        # invoke Car's constructor
        return cls(make, model, year)

    def __repr__(self): 
        instance_str = f'Car(make={self.make}, model={self.model}, year={self.year}, '
        instance_str += f'wheels={Car.wheels})'
        return instance_str

In [24]:
# now would be called this way
print(Car.which_is_newer(car1, car2))

temp_car = Car("make", "model", "2003")
print(temp_car.which_is_newer(car1, car2))

Car(make=Pontiac, model=Grand Am, year=1997, wheels=4)
Car(make=Pontiac, model=Grand Am, year=1997, wheels=4)


### Encapsulation

>``[Encapsulation] allows the implementation of an object's interface to be changed without impacting the users of that object."

The main idea of encapsulation is to hide implementation details from the users of an object. You only expose a public interface to the users.

There are a few ways to encapsulation is handled in Python: 

- Private attributes using underscores
- Getter/Setters
- Properties

### Private Attributes

We saw last week, if we define class attributes with double underscores they are not accessible outside the class.

In [25]:
class Example:
    def __init__(self, x, y, z):
        self.x = x
        self._y = y
        self.__z = z
        
    def __repr__(self):
        return f"Example({self.x}, {self._y}, {self.__z})"

instance = Example(1, 2, 3)

In [26]:
# normal public attribute
instance.x

1

In [27]:
# single underscore attributes are private by convention only
# (there is no enforcement)
instance._y

2

In [28]:
# double underscore methods are name-mangled
instance.__z

AttributeError: 'Example' object has no attribute '__z'

In [29]:
instance

Example(1, 2, 3)

### Getters / Setters

Another common pattern to hide data in OOP languages is to use getter and setter methods that control access.

In [30]:
class Person:
    def __init__(self, name, age):
        self.__name = name  #  Assume it has getter/setters 
        self.set_age(age)

    def get_age(self):
        return self.__age

    def set_age(self, age):
        if age < 0:
            raise ValueError("Person can't have a negative age!")
        self.__age = age
        
    def set_name(self, name):
        if " " not in name:
            raise ValueError("must be at least two words")
        self.__name = name

In [31]:
p = Person("Joe Biden", 79)

In [32]:
p.get_age()

79

In [33]:
p.set_age(80)

In [34]:
p.get_age()

80

In [35]:
p.set_age(-1)

ValueError: Person can't have a negative age!

In [36]:
p.get_age()

80

This can become very tedious, and as we've seen they don't actually protect access to variables.  Therefore we typically **avoid getters and setters in Python.**

### Properties

We want the advantages of encapsulation (being able to avoid improper use, hiding our internal representation, etc.) but without the need to start with a bunch of getter/setter functions from the get go.

Python has a much nicer way to control access to attributes via **properties**.

There is a built in function `property()` that creates and returns a property object.

`property(fget=None, fset=None, fdel=None, doc=None)`

- `fget` is a function to get value of the attribute
- `fset` is a function to set value of the attribute
- `fdel` is a function to delete the attribute
- `doc` is a docstring for the attribute

In [35]:
class Person:
    def __init__(self, name, age):
        self.__name = name  #  Assume it has getter/setters 
        self.age = 4

    def get_age(self):
        print("inside get age")
        return self.__age

    def set_age(self, age):
        if age < 0:
            raise ValueError("Person can't have a negative age!")
        self.__age = age
        
    def __repr__(self):
        return f"Person({self.__name!r}, {self.__age})"
        
    age = property(get_age, set_age)

In [36]:
p = Person("Wayne", 30)
p.age = 31
print(p)

Person('Wayne', 31)


In [37]:
p.age = -1

ValueError: Person can't have a negative age!

In [38]:
print(p.age)

inside get age
31


#### @property

We can also use `property` as a decorator. 

- Place the `@property` directly above the function header of the getter function.

- Place the code `@name_of_property.setter` above the function header of the setter function. You need to replace the name_of_property with the actual name of the property.

- The function names for both the setter/getter need to match.

In [37]:
class Person:
    def __init__(self, name, age):
        self.__name = name  #  Assume it has getter/setters 
        # invokes setter
        self.age = age

    @property
    def age(self):
        return self.__age
    #age = property(age)
    
    @age.setter
    def age(self, age):
        if age < 0:
            raise ValueError("Person can't have a negative age!")
        self.__age = age

    def __repr__(self):
        return f"Person({self.__name!r}, {self.__age})"

In [39]:
p2 = Person("Emma", 28)
#p2.age = -1
p2.age = 4
print(p2.age)

TypeError: 'int' object is not callable

This allows us to start class attributes as public, and add properties as needed.

In [110]:
class Point:
    def __init__(self, x, y):
        self.x = x 
        self.y = y

In [111]:
p = Point(10, 10)

#### Read-only/Calculated Properties

In [42]:
class Rectangle: 
    
    def __init__(self,width,height):
        self.width = width 
        self.height = height 
        self.other_area = self.width * self.height
    
    # read-only calculated property
    @property 
    def area(self):
        return self.width * self.height 

In [43]:
r = Rectangle(3, 9)


In [45]:
print(r.other_area, r.area)

27 27


In [46]:
r.width = 6
print(r.other_area, r.area)

27 54


In [115]:
r.area = 4

AttributeError: can't set attribute 'area'

## Inheritance

### Motivations

Let's say we're building an application that tracks students.

In [42]:
class Student:
    
    next_id_counter = 1
    
    def __init__(self, name):
        # assign each student a unique id
        self.id = Student.next_id_counter
        Student.next_id_counter += 1
        
        self.name = name
        self.year = 1
        self.major = "Undeclared"
        self.course_grades = {}
        
    def add_grade(self, course_name, grade):
        self.course_grades[course_name] = grade
    
    @property
    def gpa(self):
        grade_pts = {"A":4.0, "A-":3.7, "B+":3.3, "B":3.0, "B-":2.7, "C+":2.3, "C":2.0, "C-":1.7, "D+":1.3, "D":1.0, "F":0.0} 
        if len(self.course_grades) == 0:
            return 0
        return sum(grade_pts[g] for g in self.course_grades.values()) / len(self.course_grades)
    
    def __repr__(self):
        return f"Student(name={self.name}, id={self.id}, gpa={self.gpa})"

In [43]:
s1 = Student("Adam")
s2 = Student("Beth")
s2.add_grade("Programming Python", "A")
s2.add_grade("Discrete Math", "B+")

In [44]:
print(s1)
print(s2)

Student(name=Adam, id=1, gpa=0)
Student(name=Beth, id=2, gpa=3.65)


Perhaps we want to add `Alumni` to our application.

An alum will have some things in common with students:

- They still have a name.
- We want to remember their major.
- We'll still want to keep track of their grades/GPA.

We now also:

- Want to record their year of graduation.
- No longer want to allow grades to be recorded.
- Want to be able to calculate how long ago they graduated.
- When displaying them, we want to display their graduation year.

**How to implement?**

We *could* copy `student.py` and rename to `alum.py` and rename the class as needed.

**But copying & pasting is generally a bad idea!**

We'd need to fix bugs & add features in both classes separately.

A new feature in `Student` would need to be copied over to `Alum`, this will quickly get messy.


### Implementation

Instead we will use **inheritance**, which allows us to create a new class from an existing one.  The new class inherits the attributes and methods from the parent.

- **superclass**, **parent**, or **base** class: The pre-existing class.
- **subclass**, **child**, or **derived** class: The new class that inherits the code (attributes & methods) of another class.

Subclasses can extend/modify the functionality of superclasses.

Syntax:

```python
class Subclass(Superclass):
    pass
```

For example:

```python
class Alum(Student):
    pass
```

At this point, `Alum` is a new class with the exact same implementation as `Student`.

Typically we'll want to add new instance & class variables, methods, etc.

Newly defined features will only apply to instances of `Alum`

It is possible to override parent class behavior, or rely on parent behavior, whichever is needed.

### Adding & Overriding Behavior

In [45]:
class Alum(Student):
    def __init__(self, name, grad_year):
        # call Student's constructor, which contains id logic
        super().__init__(name)
        self.graduation_year = grad_year
        
    # new behavior
    def years_since_graduation(self, now):
        return now - self.graduation_year
    
    # overrides parent's add_grade
    def add_grade(self, course_name, grade):
        print("Sorry, you cannot add grades to Alums")
        # we choose not call super().add_grade here
    
    # overrides parent's __repr__
    def __repr__(self):
        return f"Alum(name={self.name}, id={self.id}, gpa={self.gpa}, graduated={self.graduation_year})"

In [46]:
alum1 = Alum("Charlie", 2016)
print(alum1)
print(alum1.years_since_graduation(2022), "years since graduation")
alum1.add_grade("Python", "B")
alum1.gpa

Alum(name=Charlie, id=3, gpa=0, graduated=2016)
6 years since graduation
Sorry, you cannot add grades to Alums


0

### super()

Allows direct access to parent class(es).

### issubclass & isinstance

In [168]:
isinstance(4, int)

True

In [169]:
isinstance([1, 2, 3], list)

True

In [170]:
s1 = Student("Sarah")
isinstance(s1, Student)

True

In [171]:
# child classes are instances of parent types
alum1 = Alum("Charlie", 2016)
isinstance(alum1, Student)

True

In [56]:
# but not vice-versa
isinstance(s1, Alum)

False

In [175]:
# takes class names
issubclass(Alum, Student)

True

In [57]:
issubclass(Student, Alum)

False

In [58]:
# We've been saying "everything in Python is an object"

# all of these are instances of `object`
(isinstance(4, object)
and isinstance(print, object)
and isinstance(lambda x: x ** 2, object)
and isinstance({"a": 1}, object)
and isinstance(alum1, object)
and isinstance(Student, object))

True

In [60]:
repr(lambda x: x)

'<function <lambda> at 0x1053e8940>'

### `object`

Every object derives from a base class named `object`.

```python
class Point:
    def __init__(self, x, y):
        self.x = y

# Same as:

# do not write explicitly in Python 3
class Point(object):
    def __init__(self, x, y):
        self.x = y
        self.y = y
```

### MRO

When we call a function, Python walks up the chain of parent classes to determine the first one that has the method defined.

This is called the **method resolution order**.


In [185]:
help(alum1)

Help on Alum in module __main__ object:

class Alum(Student)
 |  Alum(name, grad_year)
 |  
 |  Method resolution order:
 |      Alum
 |      Student
 |      builtins.object
 |  
 |  Methods defined here:
 |  
 |  __init__(self, name, grad_year)
 |      Initialize self.  See help(type(self)) for accurate signature.
 |  
 |  __repr__(self)
 |      Return repr(self).
 |  
 |  add_grade(self, course_name, grade)
 |      # overrides parent's add_grade
 |  
 |  years_since_graduation(self, now)
 |      # new behavior
 |  
 |  ----------------------------------------------------------------------
 |  Readonly properties inherited from Student:
 |  
 |  gpa
 |  
 |  ----------------------------------------------------------------------
 |  Data descriptors inherited from Student:
 |  
 |  __dict__
 |      dictionary for instance variables (if defined)
 |  
 |  __weakref__
 |      list of weak references to the object (if defined)
 |  
 |  ------------------------------------------------------

## Abstract Base Classes

Sometimes we want to define a class that can't be instantiated directly, but is intended to be inherited from.

These are known as **abstract classes**.  This helps us define an interface, which contains a collection of methods that the **concrete class** must implement.



In [47]:
def print_dot_prod(v1, v2):
    """ prints dot product between two vectors """
    print(v1.dot_product(v2))

If we want this  method to be polymorphic for vectors of multiple dimensions.

In [48]:
class Vec2:
    def __init__(self,x,y):
        self.x = x
        self.y = y  
        
class Vec3:
    def __init__(self,x,y,z):
        self.x = x
        self.y = y  
        self.z = z 

We can force that these types implement an interface (i.e., an abstract base class) such that we can guarantee that objects we pass to ``print_dot_prod`` will always work by forcing them to implement a ``dot_product`` method. 

We will define an abstract class called ``Vector`` that has only the required method: 

`` def dot_product(self, other) `` 

In [49]:
from abc import ABC, abstractmethod

class Vector(ABC):    
    @abstractmethod
    def dot_product(self, other):
        pass

In [50]:
# we can't instantiate abstract classes
v = Vector()

TypeError: Can't instantiate abstract class Vector with abstract method dot_product

In [51]:
class Vec2(Vector):
    def __init__(self, x, y):
        self.x = x
        self.y = y  
        
    def dot_product(self, other): 
        return self.x * other.x + self.y * other.y
        
class Vec3(Vector):
    def __init__(self, x, y, z):
        self.x = x
        self.y = y  
        self.z = z 
        
    def dot_product(self, other): 
        return self.x * other.x + self.y * other.y + self.z * other.z

In [52]:
# now print_dot_prod works

# Vec2 and Vec3 objects are instances of Vector since their classes 
# inherit from the Vector ABC.
v2a = Vec2(1,2)
v2b = Vec2(3,4)
v3a = Vec3(6,7,3)
v3b = Vec3(1,2,3)

print(isinstance(v2a, Vec2)) 
print(isinstance(v2a, Vector)) 
print("----")
print(isinstance(v3a, Vec3)) 
print(isinstance(v3a, Vector))

True
True
----
True
True


In [53]:
print_dot_prod(v2a, v2b)

11


In [54]:
print_dot_prod(v3a, v3b)

29


## Dataclasses

Python 3.7 added `dataclasses` as a handy way to create classes that are mostly responsible for representing data. These classes often have few or no methods defined.

In [55]:
from dataclasses import dataclass

@dataclass
class InventoryItem:
    """Class for keeping track of an item in inventory."""
    name: str
    unit_price: float
    quantity_on_hand: int = 0

    def total_cost(self) -> float:
        return self.unit_price * self.quantity_on_hand


In [60]:
wrench = InventoryItem("Wrench", 12.95, 10)
hammer = InventoryItem("Hammer", 16, 8)
nails = InventoryItem("Nails", 0.03, 1000)
nails2 = InventoryItem("Nails", 0.03, 1000)
saw = InventoryItem("Saw", 99)

Dataclasses get an automatic `__init__`, `__repr__`, `__eq__`, and several other helpful options.  (Even more is possible via the decorator: https://docs.python.org/3/library/dataclasses.html)

In [61]:
nails.total_cost()
print(nails == nails2)

True


Beyond this, additional methods/staticmethods/etc. can be defined in the usual way.

This syntax uses Python's type-hinting, and if you're looking to use it you'll want to get familiar with the rules
around complex types: https://docs.python.org/3/library/typing.html

Note: These rules have been evolving rapidly from Python 3.6->now.

**This is one of the areas where Python version matters, if you are using type hints in this class be sure to do so
in a way compatible with Python 3.8**

